rust/src/types/node.rs

```diff
@@ -0,0 +1,180 @@
+)]
+#[repr(C)]
+pub struct ValueWithIndex {
+    /// The index of the value in the leaf node.
+    pub index: u8,
+    /// The value stored in the leaf node.
+    pub value: Value,
+}
+
+/// A sparsely populated leaf node in a (file based) Verkle trie.
+// NOTE: This type should NOT implement [`Clone`] because there should never be two instances
+// corresponding to the same logical node.
+// NOTE: Changing the layout of this struct will break backwards compatibility of the
+// serialization format.
+#[derive(Debug, PartialEq, Eq, FromBytes, IntoBytes, Immutable)]
+#[cfg_attr(test, derive(Clone))] // For testing purposes only
+#[repr(C)]
+pub struct SparseLeafNode<const N: usize, C: Commitment> {
+    pub commitment: C,
+    pub stem: [u8; 31],
+    pub values: [ValueWithIndex; N],
+}
+
+impl<const N: usize, C> Default for SparseLeafNode<N, C>
+where
+    C: Commitment + Default,
+{
+    fn default() -> Self {
+        let mut values = [ValueWithIndex::default(); N];
+        values.iter_mut().enumerate().for_each(|(i, v)| {
+            v.index = i as u8;
+        });
+
+        SparseLeafNode {
+            commitment: C::default(),
+            stem: [0; 31],
+            values,
+        }
+    }
+}
+
+/// A leaf node with 256 children in a (file based) Verkle trie.
+// NOTE: This type should NOT implement [`Clone`] because there should never be two instances
+// corresponding to the same logical node.
+// NOTE: Changing the layout of this struct will break backwards compatibility of the
+// serialization format.
+#[derive(Debug, PartialEq, Eq, FromBytes, IntoBytes, Immutable)]
+#[cfg_attr(test, derive(Clone))] // For testing purposes only
+#[repr(C)]
+pub struct FullLeafNode<C: Commitment> {
+    pub commitment: C,
+    pub stem: [u8; 31],
+    pub values: [Value; 256],
+}
+
+impl<C> Default for FullLeafNode<C>
+where
+    C: Commitment + Default,
+{
+    fn default() -> Self {
+        FullLeafNode {
+            commitment: C::default(),
+            stem: [0; 31],
+            values: [Value::default(); 256],
+        }
+    }
+}
+
+/// An inner node in a (file based) Verkle trie.
+// NOTE: This type should NOT implement [`Clone`] because there is never be two instances
+// corresponding to the same logical node.
+// NOTE: Changing the layout of this struct will break backwards compatibility of the
+// serialization format.
+#[derive(Debug, PartialEq, Eq, FromBytes, IntoBytes, Immutable, Unaligned)]
+#[cfg_attr(test, derive(Clone))] // For testing purposes only
+#[repr(C)]
+pub struct InnerNode<C: Commitment> {
+    pub commitment: C,
+    pub values: [NodeId; 256],
+}
+
+impl<C> Default for InnerNode<C>
+where
+    C: Commitment + Default,
+{
+    fn default() -> Self {
+        InnerNode {
+            commitment: C::default(),
+            values: [NodeId::from_idx_and_node_type(0, NodeType::Empty); 256],
+        }
+    }
+}
+
+/// A node in a (file based) Verkle trie.
```

Comment:
```suggestion
/// A node in a (file based) Verkle trie.
/// Non-empty nodes are stored as boxed to save on memory (otherwise the size of [Node] would be dictated by the largest variant).
```